### PR TITLE
[CCR] Add index setting to keep track of the index uuid of the leader index in the follow index

### LINKED ISSUE #31505: [CCR] Introduce leader index setting for follow index
Associate the uuid of the leader index in the follow index's index settings. 
The follow index api should fail if this index setting in the follow index and
should check if the uuid in the follow index matches with uuid of the leader index.

This validation will prevent a follow index from following an incompatible leader index.

The create and follow api should introduce this setting when it creates the follow index.

Relates to #30086

### PR DESCRIPTION
**Pull request by martijnvg**
The follow index api checks if the uuid in the follow index matches
with uuid of the leader index and fails otherwise. This validation will
prevent a follow index from following an incompatible leader index.

The create_and_follow api will automatically add this index setting
when it creates the follow index.

Closes #31505

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrSettings.java [hunk 1]
```diff
@@ -32,6 +32,12 @@ private CcrSettings() {
     public static final Setting<Boolean> CCR_FOLLOWING_INDEX_SETTING =
             Setting.boolSetting("index.xpack.ccr.following_index", false, Setting.Property.IndexScope);
 
+    /**
+     * An internal setting used to keep track to what leader index a follow index associated. Do not use externally!
```
Comment (jasontedor) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrSettings.java]: I think that we need a different mechanism here. Either we need to invent the notion of a private setting, or as I prefer, these should be considered custom index metadata. While we have partially deprecated and neutered custom index metadata, I think that we should bring it back. We also have a similar need in index lifecycle management. As @colings86, @talevy, and me are all out next week, I will pick up a thread when I return from vacation to resuscitate custom index metadata and then we can use it here and in index lifecycle management.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a comment that I think we should use custom index metadata here.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrSettings.java (hunk 1 above)
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrSettings.java]: Thanks @jasontedor for sharing your thoughts on this setting. When I added the setting, it felt wrong to me that this was a public setting, hence the comment. I think your idea to use custom index metadata makes sense here.

**Comment by jasontedor** [comment]
@martijnvg I think since we are going to take a different approach, we can close this PR and open a new one based on the index metadata customs?

**Comment by martijnvg** [comment]
@jasontedor Yes, I will open a new PR.

**Pull request closed without merging**